<a href="https://colab.research.google.com/github/Luisafe-ux/Business-Intelligence./blob/main/RETO1_NaiveBayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

0. Se procede a cargar las librerías de trabajo

In [10]:
import numpy as np
import pandas as pd


#conectar al Drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


1. Cargar la base de datos

In [11]:
nxl='/content/drive/MyDrive/1. BD1_Longitudinal Survey_Int.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)
XDB.head(50)

#Se seleccionan las variables de interés
XDB=XDB[['Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income','Urban']]
XDB.head(10)

columnas_considerar = ['Age', 'Mother_Edu', 'Father_Edu', 'Height', 'Weight', 'Income', 'Self_Esteem', 'FamilySize', 'Urban']
XDB_1 = XDB.dropna(subset=columnas_considerar)
XDB_1.head(50)

#Se separan las variables de entrada y salida
XD=XDB_1[['Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income']] #Variables de Entrada
yd=XDB_1[['Urban']]   #Variable de salida - variable Pronóstico

2. Número total de la base de datos antes y después de la depuración

In [12]:
# Obtener el número total de datos antes de la depuración
total_datos = len(XDB)
print(f"Número total de datos: {total_datos}")

Número total de datos: 12686


In [13]:
# Obtener el número total de datos después de la depuración
total_datos1 = len(XDB_1)
print(f"Número total de datos: {total_datos1}")

Número total de datos: 5948


3. Porcentaje de los datos que hacen parte de cada una de las categorías de clasificación antes y después de la depuración.

In [ ]:
# Obtener el porcentaje de datos en las categorías de clasificación antes de la depuración

porcentaje= (XDB.count() / total_datos) * 100
print("\nPorcentaje de datos por variable:")
print(porcentaje)




Porcentaje de datos por variable:
Age            100.000000
Mother_Edu      93.630774
Father_Edu      85.763834
FamilySize     100.000000
Self_Esteem     94.529402
Height          95.719691
Weight          95.632981
Income          59.703610
Urban           99.684692
dtype: float64


In [ ]:
# Obtener el porcentaje de datos en las categorías de clasificación antes de la depuración

porcentaje= (XDB_1.count() / total_datos1) * 100
print("\nPorcentaje de datos por variable:")
print(porcentaje)



Porcentaje de datos por variable:
Age            100.0
Mother_Edu     100.0
Father_Edu     100.0
FamilySize     100.0
Self_Esteem    100.0
Height         100.0
Weight         100.0
Income         100.0
Urban          100.0
dtype: float64


In [ ]:
# Obtener el porcentaje de los que viven en el área urbana y los que viven en área rural.
ND1=len(np.where((yd==1))[0])
print(ND1/len(XD))      #vive en el área urbana

ND2=len(np.where((yd==0))[0])
print(ND2/len(XD))      #no vive en el área urbana


0.7916946872898454
0.20830531271015468


3. Determinar el intervalo en el cual se encuentra la variación de las siguientes variables: Edad, Altura y Peso.

In [ ]:
from sklearn.naive_bayes import GaussianNB

mnb=GaussianNB()
mnb.fit(XD,yd)

#Obtenemos la información del modelo
u=mnb.theta_
print("la media de 'Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income es:\n",u)
var=mnb.var_  #Se calculan las varianzas    indica que tan alejados están los datos de la media
sigma=np.sqrt(var) #Se calcula la deviación
print("la desviación de  'Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income es :\n",sigma)


#Se determina los limites superiores e inferiores de los ingresos
LS= u+sigma
LI= u-sigma
print("El limite superior de  'Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income es:\n",LS)
print("El limite inferior de  'Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income es:\n",LI)


la media de 'Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income es:
 [[1.83640032e+01 1.07312349e+01 1.02711864e+01 4.73930589e+00
  2.19951574e+01 6.72429379e+01 1.49003228e+02 3.01248854e+04]
 [1.83618603e+01 1.11613931e+01 1.12331705e+01 4.68103631e+00
  2.25744319e+01 6.69116585e+01 1.44750903e+02 3.32736742e+04]]
la desviación de  'Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income es :
 [[2.47260545e+00 3.15661127e+00 3.88837565e+00 2.55226748e+00
  4.21650807e+00 4.08228636e+00 3.03486350e+01 3.16482374e+04]
 [2.47344359e+00 3.40066236e+00 4.11549881e+00 2.44205851e+00
  4.19763081e+00 4.20193102e+00 3.00285301e+01 3.60897302e+04]]
El limite superior de  'Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income es:
 [[2.08366087e+01 1.38878461e+01 1.41595621e+01 7.29157338e+00
  2.62116655e+01 7.13252242e+01 1.79351863e+02 6.17731228e+04]
 [2.08353039e+01 1.45620554e+01 1.53486693e+01 7.

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


4. Evaluación de un registro

In [ ]:
# Define las variables de entrada y salida
#Se separan las variables de entrada y salida
XD=XDB_1[['Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income']] #Variables de Entrada
yd=XDB_1[['Urban']]   #Variable de salida - variable Pronóstico

# Convierte yd en un array 1D
yd = yd.values.ravel()

# Crea y entrena el modelo
mnb = GaussianNB()
mnb.fit(XD, yd)

# Evaluación a un habitante
XDS = [[34, 12, 15, 2, 24, 69, 160, 48000]]

# Realiza la predicción
ydp = mnb.predict(XDS)

# Imprime los resultados
print("La predicción del habitante es:", ydp[0])  # Accede al primer elemento de ydp
if ydp==0:
  print("El habitante no vive en zona urbana")

if ydp==1:
  print("El habitante vive en zona urbana")


La predicción del habitante es: 1.0
El habitante vive en zona urbana


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


5. Evaluación de desempeño del modelo.




In [ ]:
from functools import cmp_to_key
from sklearn.metrics import confusion_matrix

#Haga un barrido de los datos
ydp2=mnb.predict(XD) #Haga el pronostico para los datos base
cm=confusion_matrix(yd,ydp2) #Cree la matriz de confusión
print("La matriz de confusión es:\n",cm)
VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]

#METRICAS DESEMPEÑO
Ex=(VP+VN)/(VP+VN+FP+FN)
print("La exactitud del modelo es:\n",Ex) #Exactitud: comportamiento general preaprobando

TE=(FN+FP)/(VP+VN+FP+FN)
print("La tasa de error del modelo es:\n",TE) #Tasa de error: Comportamiento general no preaprobado

Se=VP/(VP+FN)
print("La sensibilidad del modelo es:\n",Se) #Sensibilidad: comportamiento correcto preaprobando

Sp=VN/(VN+FP)
print("La especificidad del modelo es:\n",Sp) #Especificidad:comportamiento correcto no preaprobando

pre=VP/(VP+FP)
print("La precisión del modelo es:\n",pre) #Precisión: comportamiento correcto preaprobando

PrNeg=VN/(VN+FN)
print("La probabilidad negativa del modelo es:\n",PrNeg) #Probabilidad negativa: comportamiento correcto no preaprobando

FP=cm[0,1]
FN=cm[1,0]
VP=cm[1,1]

La matriz de confusión es:
 [[   3 1236]
 [   8 4701]]
La exactitud del modelo es:
 0.7908540685944856
La tasa de error del modelo es:
 0.20914593140551446
La sensibilidad del modelo es:
 0.9983011255043533
La especificidad del modelo es:
 0.002421307506053269
La precisión del modelo es:
 0.7918140474987367
La probabilidad negativa del modelo es:
 0.2727272727272727


**ANÁLISIS DE RESULTADOS**

De acuerdo con la base de datos suministrada, se tomaron como referencia para determinar si una persona vive en un área urbana o en un área rural las variables edad, educación de los padres, estatura, peso, ingresos, escala de autoestima y tamaño de la familia. La base de datos posee un total de 5948 registros, donde el (79,16%) corresponde a los habitantes que viven en área urbana y el (20,83%) corresponde a los que viven en área rural. De manera general, el modelo alcanzó una exactitud del (79,08%), lo que indica un buen desempeño en la clasificación de los datos correspondientes a los habitantes que viven en el área urbana. Sin embargo, muestra una tasa de error del (20,91%), lo que sugiere que aproximadamente una de cada cinco clasificaciones es incorrecta. Con respecto a la métrica de sensibilidad, el modelo tiene un valor de (99,83%), lo que significa que identifica de manera casi perfecta a los habitantes de áreas urbanas. Sin embargo, la especificidad del modelo es extremadamente baja (0,24%), lo que indica que tiene dificultades para identificar correctamente a los habitantes de zonas rurales. Por otro lado, tenemos que la precisión del modelo es de (79,18%), lo que muestra que la mayoría de las predicciones positivas clasificadas como urbanas son correctas. Sin embargo, la baja especificidad sugiere que el modelo clasifica erróneamente a muchos habitantes rurales como urbanos. Finalmente, la probabilidad negativa del modelo es de (27,27%,) lo que indica que existe una proporción considerable de habitantes rurales que no están siendo reconocidos correctamente. Lo anterior nos muestra que aunque el modelo tiene una buena capacidad para identificar habitantes urbanos, su desempeño en la clasificación de habitantes rurales es deficiente.
